treeherder/log_parser/crossreference.py

```diff
@@ -0,0 +1,117 @@
+import logging
+import re
+
+from mozlog.formatters.tbplformatter import TbplFormatter
+
+from treeherder.model.models import (FailureLine,
+                                     TextLogError,
+                                     TextLogSummary,
+                                     TextLogSummaryLine)
+
+logger = logging.getLogger(__name__)
+
+
+def crossreference_job(job):
+    """Populate the TextLogSummary and TextLogSummaryLine tables for a
+    job. Specifically this function tries to match the
+    unstructured error lines with the corresponding structured error lines, relying on
+    the fact that serialization of mozlog (and hence errorsummary files) is determinisic
+    so we can reserialize each structured error line and perform an in-order textual
+    match.
+
+    :job: - Job for which to perform the crossreferencing
+    """
+    failure_lines = FailureLine.objects.filter(job_guid=job.guid)
+
+    text_log_errors = TextLogError.objects.filter(
+        step__job__guid=job.guid).order_by('line_number')
```

Comment:
You can just do `step__job=job` here.

Incidentally it would be nice if FailureLine and friends referenced the Job model directly via a foreign key relation, but that can obviously wait for another day.